### PR TITLE
Add some Rust building infrastructure and example

### PR DESCRIPTION
**Pull request by chrysn**
### Contribution description

While writing Rust applications has been possible for some time, it has required a few Makefile rules which I'm by now happy enough with to start a first PR. Having this all in-tree is not essential for operation, but makes it a bit easier to store, for example, the target definitions with the architectures rather than [in a big if chain](https://gitlab.com/etonomy/riot-examples/-/blob/master/Makefile.cargo-target) that sometimes peeks into fields and sometimes into boards.

This adds
* Rust target definitions to platforms -- basically Rust's version of a platform triple -- in the architectures
* Build rules to create something almost, bot not quite, entirely unlike modules -- it's called `$(APPLICATION_RUST_MODULE).module` and added to `BASELIBS`, but is built by running Cargo and then exploding the built static lib into the folder where LINK expects it to be.
* Two examples

### Testing procedure

* Ensure you have Rust installed -- given you need nightly versions, typically by installing [rustup](https://rustup.rs/)
  * Alternatively, run with `BUILD_IN_DOCKER=1`
* In examples/rust-hello-world (or rust-gcoap), run `make all flash term`, optionally with a cortex M0-4 board.

It is tested so far with native and some cortex-m0, 3 and 4 boards (mostly EFM32, EFR32 and nRF52840). riscv support is disabled right now, I don't want to do that in the first rush until it falls into place trivially.

### Questions:

* Did I get the what-goes-where of the Makefiles right?
* Would it make more sense to build something parallel to `${MODULE}.module` that is `${CRATE}.crate` that gets its own place in the big find of LINK? I think it'd make sense because it wouldn't involve overriding a target any more, and moreover make it easier to later extend this to having potentially multiple built crates (although I think it's preferable to use one), but would make a larger diff.
  * If that's solved, I can probably come up with a nicer way of saying, in the example, that "this has a crate as its main rather than a module", and show less guts in there.

### Issues/PRs references

Closes: https://github.com/RIOT-OS/RIOT/issues/9799

[edit: Updated examples list and test description now that the riotdocker images just do the right thing]

### CODE REVIEW

**Comment by chrysn** [comment]
Oh, one addendum to testing: GCC is not really well supported as a toolchain; if things break, try with TOOLCHAIN=llvm. With https://github.com/RIOT-OS/RIOT/pull/16129 merged, GCC support should increase because @maribu did a much better job in finding the subtle differences than I did in the workarounds currently in place in the riot-sys crate.

**Comment by chrysn** [comment]
Note that using this branch currently breaks most of the riot-examples as they use the older "linking in a .a archive" approach and not the newer "unar everything and let find find the .o files" one.

The shell example already uses the new style and works; I expect to port over the others while this PR matures.

**Comment by chrysn** [comment]
There's a follow-up branch at https://github.com/chrysn-pull-requests/RIOT/tree/rust-lib in which a module is introduced that builds any number of (predefined but pseudomodule-enabled) crates in a single build and pulls them in. By the power of XFA, thus pure-Rust modules for RIOT can be implemented without any C code as long as they hook into XFA extension points (or, probably, interrupts).

**Comment by chrysn** [comment]
For this to be runnable and testable in CI, https://github.com/RIOT-OS/riotdocker/pull/141 will need to be resolved.

**Comment by chrysn** [comment]
This can now be tested also using a branch of the riotdocker image; testing procedure has been amended.

**Comment by benpicco** [comment]
I have heard the CI is updated now - this needs a rebase.

**Comment by kaspar030** [comment]
> I have heard the CI is updated now - this needs a rebase.

no it is not yet!

File: makefiles/arch/cortexm.inc.mk
Comment (kaspar030) [inline makefiles/arch/cortexm.inc.mk:115]:
```suggestion
  # This is OK while Rust is not being used; if it is, the absence of a
```

**Comment by chrysn** [comment]
Squashed and rebased after having become a bit stale.

Just running the #16109 Makefile.ci generation locally (twice now, as before I didn't have all Rust targets installed), and then I'll push the fancy "CI: Ready to build" button.

**Comment by chrysn** [comment]
The Cortex M0 boards are still a bit troublesome, and are disabled for the time being. (Some seem to work with TOOLCHAIN=llvm). Same for RISC-V, but left on to see if builds fail in CI the same way. No clue what upsets things on the mulle board.

Let's give it a spin...

**Comment by chrysn** [comment]
OK, finally I see new errors pop up from CI, this is what I hoped for :-)

Examples:

[cc2650stk](https://ci.riot-os.org/RIOT-OS/RIOT/16274/6abd413a965837f6db470a8cb56cc22d6e84a22a/output/compile/examples/rust-hello-world/cc2650stk:gnu.txt) and [limifrog-v1](https://ci.riot-os.org/RIOT-OS/RIOT/16274/6abd413a965837f6db470a8cb56cc22d6e84a22a/output/compile/examples/rust-hello-world/limifrog-v1:gnu.txt)

```
-- running on worker riotci-worker-icc-autoscaled-7c7947d86c-n5qww thread 1, build number 5781.
make: Entering directory '/tmp/dwq.0.35874901328774356/9b4614b84f7f892aeb77e76ef0fdd008/examples/rust-hello-world'
Building application "hello-world" for "cc2650stk" with MCU "cc26x0_cc13x0".

   Compiling riot-sys v0.4.2
   Compiling riot-wrappers v0.7.6
   Compiling rust-hello-world v0.1.0 (/tmp/dwq.0.35874901328774356/9b4614b84f7f892aeb77e76ef0fdd008/examples/rust-hello-world)
    Finished release [optimized] target(s) in 7.66s
arm-none-eabi-ar: /tmp/dwq.0.35874901328774356/9b4614b84f7f892aeb77e76ef0fdd008/target/thumbv7m-none-eabi/release/librust_hello_world.a: No such file or directory
make: *** [/tmp/dwq.0.35874901328774356/9b4614b84f7f892aeb77e76ef0fdd008/makefiles/cargo-targets.inc.mk:27: rust_hello_world.module] Error 9
make: Leaving directory '/tmp/dwq.0.35874901328774356/9b4614b84f7f892aeb77e76ef0fdd008/examples/rust-hello-world'
cat: /tmp/dwq.0.35874901328774356/9b4614b84f7f892aeb77e76ef0fdd008/build/test-input-hash.sha1: No such file or directory
-- build directory size: 1020K
```

weird, it built successfully but then didn't find the .a? Are any weird options set, maybe?

[hifive1b](https://ci.riot-os.org/RIOT-OS/RIOT/16274/6abd413a965837f6db470a8cb56cc22d6e84a22a/output/compile/examples/rust-hello-world/hifive1b:gnu.txt)

```
  --- stderr
  thread 'main' panicked at 'libclang error; possible causes include:
  - Invalid flag syntax
  - Unrecognized flags
  - Invalid flag arguments
  - File I/O errors
  - Host vs. target architecture mismatch
  If you encounter an error missing from this list, please file an issue or a PR!', /data/riotbuild/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.53.3/src/ir/context.rs:573:15
```

This could be fixed with https://github.com/RIOT-OS/riotdocker/pull/153 as that aligns versions of llvm.

**Comment by chrysn** [comment]
CI green :-D (OK: Murdock green. And a christmas tree of red lights on the rest.)

Next steps for me are to lock down crate versions, squash this into plausible steps (it has accumulated some back-and-forth again, and some fixups are not trivially assignable), and then send a ping for reviews while the "limit tests to Rust and some other test" plug is still on. Once reviews are OK, the plug can go off, CI runs through if it so be, and off we can go, hopefully.

**Comment by chrysn** [comment]
Squashed into better manageable units. CARGO_TARGET renamed to RUST_TARGET (as that's what it's about; fun rebasing to keep things sorted). Crate versions are now fixed (so no matter how third parties like that strange person maintaining the riot-sys crate screws up, our builds don't suddenly start failing).

**Comment by chrysn** [comment]
I think this is now ready for review again; added @maribu because it touches compile_commands (two commits) and replicates some compile_commands infrastructure (which I think is justified but you may have a better idea there).

**Comment by chrysn** [comment]
The latest updates enable Cortex-M0 and M0+ again, added a test (so that at least some Rust code is actually run and not just built), some updates to make the examples prettier, and ample READMEs both to the examples and to the test, where future RIOT developers frustrated by the Rust tests making things complicated find hints and helpful comments.

**Comment by chrysn** [comment]
To anyone reading the history and wondering about the on and off on ready-to-build recently: There were random (~2/3 times) failures to build, which all were around disk space exhaustion.

The latest commits fix this by placing Rust's target directory inside `${BINDIR}`, but given the erratic nature of the earlier failures, still need repeated testing to be sure this didn't just get lucky.

**Comment by chrysn** [comment]
This is as ready as I think it'll get if I keep brooding on it, please consider reviewing.

I think that the toughest to review part will be makefiles ... no clue who'd be a subsystem maintainer for that to ask for a high-level check there.

**Comment by chrysn** [comment]
This is as ready as I think it'll get if I keep brooding on it, please consider reviewing.

I think that the toughest to review part will be makefiles ... no clue who'd be a subsystem maintainer for that to ask for a high-level check there.

File: cpu/cortexm_common/Makefile.features
Comment (benpicco) [inline cpu/cortexm_common/Makefile.features:36]: Is there no `thumbv8m-none-eabi`?

File: examples/rust-gcoap/Cargo.lock
Comment (benpicco) [inline examples/rust-gcoap/Cargo.lock:2]: should this be committed then?

File: examples/rust-gcoap/vfs.c
Comment (benpicco) [inline examples/rust-gcoap/vfs.c:21]:
```suggestion
#define LARGE  "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
               "1234567890---------\n"
```

File: cpu/riscv_common/Kconfig
Comment (benpicco) [inline cpu/riscv_common/Kconfig:15]: Is this currently broken?

**Review by benpicco — COMMENTED** [review]
This looks surprisingly non-scary.

File: cpu/cortexm_common/Makefile.features
Comment (chrysn) [reply cpu/cortexm_common/Makefile.features:36]: There probably is, but I haven't verified it yet. As with other architectures, I'm trying to start a bit slowly here (but if anyone has a board around where they can test the build results, I'm happy to open it up further).

File: examples/rust-gcoap/Cargo.lock
Comment (chrysn) [reply examples/rust-gcoap/Cargo.lock:2]: Yes: This is what ensures that things don't break just because any third party in the dependency chain declares an update to be compatible when it actually isn't.

Think of this file as the hashes of several pkgs that are all managed in here rolled into one. Like with packages, they'll be more or less regularly updated, but manually. (Typically, all a Cargo.lock file is updated in one go, so whenever riot-sys or -wrappers grow a feature needed for an example, everything else gets updated too).

File: cpu/riscv_common/Kconfig
Comment (chrysn) [reply cpu/riscv_common/Kconfig:15]: Yes -- or maybe not even fully, but as with the ARMv8, I'm focusing on some to get a stable reference shape of things.

**Comment by chrysn** [comment]
> surprisingly non-scary

Thanks :-)

It certainly helps that this has been growing in a separate repo, where alterations to the build process were even trickier than in-tree.

File: cpu/cortexm_common/Makefile.features
Comment (benpicco) [reply cpu/cortexm_common/Makefile.features:36]: I can test on a `saml10-xpro` which has a Cortex-M23

Comment (chrysn) [reply cpu/cortexm_common/Makefile.features:37]: I'ved added them now (and rebuilt Makefile.ci with consideration for them), but left them commented out because the build servers will need [another update round](https://github.com/RIOT-OS/riotdocker/pull/164) for them. Things work up to firmware building (which is generally a good sign, as the linker usually refuses to work if the triples don't match), and you can test it by removing the comment mark, but I'd leave them out of this PR to not make it depend on the manual CI upgrades that can only start once the above-mentioned PR is merged.

**Comment by chrysn** [comment]
Brief side note on why so many boards are listed as "doesn't fit" for the gcoap example: No measures are yet being taken to minimize; LTO and opt-level="s" would shave roughly 30% off the application size. In general I didn't tweak much yet, leaving that for later tuning steps. (For example, Rust's standard string formatting is used, and the CBOR encoding and decoding in the examples uses a very generic framework).

**Comment by chrysn** [comment]
I think that all open concerns were addressed, if only to say "out of scope for first run". (Some more Cortex M series now have triples too, but were left disabled until CI is updated all through again, and the scope point still holds).

Is this good to squash?

**Comment by benpicco** [comment]
Yes, go ahead and squash

**Review by maribu — COMMENTED** [review]
looks good to me :-)

File: cpu/native/Makefile.include
Comment (chrysn) [reply cpu/native/Makefile.include]: I've added a check in a3b7cbcf42f9b321cfaa7b600aaa7d37e757d51b to Kconfig, the .include and the .features that Rust support is only advertised for x86_64 Linux. Like with the ARM board targets, I'd keep it to what I can test in the first run, and adding more platforms like 32 bit native ARM or Darwin support (is that or is that not supported any more these days?) can be done when there are actual users who'd test it.

File: cpu/native/Makefile.include
Comment (benpicco) [inline cpu/native/Makefile.include:16]: no `arm-unknown-linux-gnu` for raspberry pi?

**Comment by chrysn** [comment]
Probably yes. But CI doesn't test it, I can't test it, and it's easy to
add if needed and and when it can be tested.

(Currently still fixing a few things during rebasing, just in case you'd
be starting another review round).

**Comment by chrysn** [comment]
Some changes during rebasing:
* Cargo.lock are not ignored any more (I had already switched to versioning them as things matured, as they're equivalent to pkg version hashes; having them in .gitignore too is allowed in git but probably not good practice)
* Fixed the precise check for native being x86_64 Linux (which failed during the CI run of the unsquashed tree because I forgot to export the OS_ARCH used for that check)
* Ran a cargo update as the newest riot-wrappers contain adjustments to https://github.com/RIOT-OS/RIOT/pull/17275 that was not breaking on C but on Rust bindings.
* Some non-fixup commits got squashed into what I now consider a manageable patch set of meaningful standalone commits

Please review, it's good to go from my side (especially, the REMOVEME commit that only makes it build the Rust examples has been removed, so it's good to ACK-and-merge if reviewers agree).

**Comment by chrysn** [comment]
The Murdock run showed two errors -- one appears to be just a random timeout (tests/pkg_edhoc_c/esp32-wroom-32), the other is highly weird because it's using an old version of a Rust library that was already updated and pinned (but maybe too softly) to the newer version. This might have been a caching issue; I'll look at it but unless this persists I don't think it's a reason for delay.

Holding off the murdock rerun that I recommend to do due to the first error until there is either an ACK or any comment that requires further changes.

**Comment by chrysn** [comment]
Correction: Nothing weird about the version thing, I just only fixed the examples and forgot about the test.

Still disabled CI rebuilding (as the fix is very likely correct, and can just as well be tested after being smushed).

I'd smush and turn on Murdock again once anyone says "ACK, smush".

**Comment by benpicco** [comment]
Feel free to squash, still looks good from my side.

**Review by maribu — APPROVED** [review]
Let me throw in an ACK. The changes look good to me and the author did rigorous testing.

**Pull request merged**
